### PR TITLE
[FX] get the correct error message

### PR DESCRIPTION
**Pull request by Chillee**
Currently, code like
```
class Test(nn.Module):
    def __init__(self):
        super(Test, self).__init__()
        self.W = torch.nn.Parameter(torch.randn(5))

    def forward(self, x):
        return torch.dot(self.W, x)

mod = Test()
print(fx.symbolic_trace(Test())(5))
```
gives an error like the below, which does not show the actual code that throws the error.
```
Traceback (most recent call last):
  File "t.py", line 20, in <module>
    print(fx.symbolic_trace(Test())(5))
  File "/home/chilli/fb/pytorch/torch/nn/modules/module.py", line 744, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/chilli/fb/pytorch/torch/fx/graph_module.py", line 191, in debug_forward
    return src_forward(self, *args, **kwargs)
  File "<eval_with_key_0>", line 5, in forward
TypeError: dot(): argument 'tensor' (position 2) must be Tensor, not int
```

This is particularly annoying when your function has already been transformed several times.

So, the really annoying thing is that the error clearly has the requisite information in `exception.__traceback__` - it just isn't printing it.

I think the right way of doing this is simply replacing `sys.excepthook`. This appears to be the standard way to modify exception messages.

**Scratch the below**

The 2 methods in the PR right now are:
1. Just prepend the final part of the traceback to the beginning of your error message. Looks like
```
Traceback (most recent call last):
  File "t.py", line 20, in <module>
    print(fx.symbolic_trace(Test())(5))
  File "/home/chilli/fb/pytorch/torch/nn/modules/module.py", line 744, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/chilli/fb/pytorch/torch/fx/graph_module.py", line 197, in debug_forward
    raise e
  File "/home/chilli/fb/pytorch/torch/fx/graph_module.py", line 192, in debug_forward
    return src_forward(self, *args, **kwargs)
  File "<eval_with_key_0>", line 5, in forward
TypeError:   File "<eval_with_key_0>", line 5, in forward
    dot_1 = torch.dot(w, x)
dot(): argument 'tensor' (position 2) must be Tensor, not int
```

2. Use the `from exception` feature in Python. Looks like
```
Traceback (most recent call last):
  File "/home/chilli/fb/pytorch/torch/fx/graph_module.py", line 192, in debug_forward
    return src_forward(self, *args, **kwargs)
  File "<eval_with_key_0>", line 5, in forward
TypeError:   File "<eval_with_key_0>", line 5, in forward
    dot_1 = torch.dot(w, x)
dot(): argument 'tensor' (position 2) must be Tensor, not int

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "t.py", line 20, in <module>
    print(fx.symbolic_trace(Test())(5))
  File "/home/chilli/fb/pytorch/torch/nn/modules/module.py", line 744, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/chilli/fb/pytorch/torch/fx/graph_module.py", line 197, in debug_forward
    raise Exception(last_tb) from e
Exception:   File "<eval_with_key_0>", line 5, in forward
    dot_1 = torch.dot(w, x)
```

I think the first one looks better, but it's pretty hacky since we're shoving the traceback in the message.

### CODE REVIEW

**Comment by jamesr66a** [comment]
Can you dig into why the default exception handling doesn't print out the last frame? Might be worth digging into some cpython source to figure out exactly what's wrong and see if we can solve it directly at the source

**Comment by Chillee** [comment]
Yeah I'll take a look. I suspect that it's by design. We're already doing some hacks on `linecache` to add the source code to the exec.

**Comment by Chillee** [comment]
@jamesr66a I think the right way to do is it overload `sys.excepthook`. This seems to be the standard way of modifying how your exception gets printed out, and doesn't require much of a change.

I'm not entirely sure what CPython source code is actually causing this, but I did find this: https://github.com/python/cpython/blob/fb5db7ec58624cab0797b4050735be865d380823/Lib/code.py#L134

I assume that somewhere they're doing something similar.

File: torch/fx/graph_module.py [hunk 1]
```diff
@@ -185,6 +187,11 @@ def recompile(self) -> None:
         """
         self.code = self._graph.python_code(root_module='self')
         cls = type(self)
+
+        def print_full_traceback(exctype, value, tb):
+            traceback.print_exception(exctype, value, tb)
+        sys.excepthook = print_full_traceback
```
Comment (jamesr66a) [inline torch/fx/graph_module.py:193]: Is this overriding the exception hook for the remainder of the lifetime of the program? If so, can we instead wrap `forward` with a function that sets this at the beginning and resets it to its original value before returning?

Comment (Chillee) [reply torch/fx/graph_module.py]: Done.

File: torch/fx/graph_module.py [hunk 2]
```diff
@@ -185,8 +187,21 @@ def recompile(self) -> None:
...
 
+        cls_call = cls.__call__
+
+        def print_full_traceback(exctype, value, tb):
+            traceback.print_exception(exctype, value, tb)
+
+        def wrapped_call(self, *args, **kwargs):
+            sys.excepthook = print_full_traceback
+            out = cls_call(self, *args, **kwargs)
+            sys.excepthook = sys.__excepthook__
```
Comment (jamesr66a) [inline torch/fx/graph_module.py:201]: How about:

```
            old_excepthook = sys.excepthook
            try:
                sys.excepthook = print_full_traceback
                return cls_call(self, *args, **kwargs)
            finally:
                sys.excepthook = old_excepthook
```

To 1) preserve another excepthook if another one was installed and 2) Make sure we reset things if the forward() call throws

File: torch/fx/graph_module.py [hunk 3]
```diff
@@ -185,8 +187,21 @@ def recompile(self) -> None:
         """
         self.code = self._graph.python_code(root_module='self')
         cls = type(self)
+
```
Comment (jamesr66a) [inline torch/fx/graph_module.py:190]: y

**Review by jamesr66a — APPROVED** [review]
Seems reasonable, I had a recommendation for improvement, but other than that this is good to go if the tests pass

**Pull request closed without merging**